stdlib/camlinternalLazy.ml

```diff
@@ -53,31 +53,40 @@ let do_force_val_block blk =
   make_forward b (Obj.repr result);
   result
 
-type status = Racy | Forcing
+type 'a status = Racy | Forcing | Done of 'a
 
-(* Assumes [blk] is a block with tag lazy *)
+(* [blk] might either be a lazy block with any tag (due to parallelism) or a
+ * short-circuited object (when the poll point inserted at the entry to this
+ * function performs a GC, which short-circuits the lazy). Hence, we need to
+ * repeat the tag checks again. *)
 let update_tag_forcing (blk : 'arg lazy_t) =
   let b = Obj.repr blk in
-  if not (update_tag b Obj.lazy_tag Obj.forcing_tag) then
-    (* blk has tag either
+  let t = Obj.tag b in
+  if t <> Obj.lazy_tag && t <> Obj.forcing_tag && t <> Obj.forward_tag then
+    Done (Obj.obj b : 'arg)
+  else if not (update_tag b Obj.lazy_tag Obj.forcing_tag) then begin
```

Comment:
Surely we would have the same bug if the safepoint gets inserted just before the `if not ...` test?